File: demos/jans-chip/ios/Jans-Chip/Models/Networking/AuthenticatorGetAssertionOptions.swift
//
//  AuthenticatorGetAssertionOptions.swift
//  Jans-Chip
//
//  Created by Nazar Yavornytskyi on 22.07.2024.
//

import Foundation

public struct AuthenticatorGetAssertionOptions: Codable {
    
    private enum CodingKeys: String, CodingKey {
        case rpId
        case clientDataHash
        case allowCredentialDescriptorList
        case requireUserPresence
        case requireUserVerification
    }
    
    var rpId: String?
    var clientDataHash: Data?
    var allowCredentialDescriptorList: [PublicKeyCredentialDescriptor]?
    var requireUserPresence: Bool?
    var requireUserVerification: Bool?
    
    func areWellFormed() -> Bool {
        if let rpId, rpId.isEmpty {
            return false
        }
                
        if clientDataHash?.count != 32 {
            return false
        }
        
        if !(requireUserPresence == true || requireUserVerification == true) { // only one may be set
            return false
        }
        
        return true
    }
}

File: demos/jans-chip/android/webauthn/src/main/java/io/jans/webauthn/models/AuthenticatorGetAssertionOptions.java
package io.jans.webauthn.models;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.SerializedName;

import java.util.List;

import io.jans.webauthn.util.Base64ByteArrayAdapter;

public class AuthenticatorGetAssertionOptions {
    @SerializedName("rpId")
    public String rpId;
    @SerializedName("clientDataHash")
    public byte[] clientDataHash;
    @SerializedName("allowCredentialDescriptorList")
    public List<PublicKeyCredentialDescriptor> allowCredentialDescriptorList;
    @SerializedName("requireUserPresence")
    public boolean requireUserPresence;
    @SerializedName("requireUserVerification")
    public boolean requireUserVerification;
    // TODO: authenticatorExtensions

    public boolean areWellFormed() {
        if (rpId.isEmpty()) {
            return false;
        }
        if (clientDataHash.length != 32) {
            return false;
        }
        if (!(requireUserPresence ^ requireUserVerification)) { // only one may be set
            return false;
        }
        return true;
    }

    public static AuthenticatorGetAssertionOptions fromJSON(String json) {
        Gson gson = new GsonBuilder()
                .registerTypeAdapter(byte[].class, new Base64ByteArrayAdapter())
                .create();
        return gson.fromJson(json, AuthenticatorGetAssertionOptions.class);
    }
}
